kubernetes/keycloak.yaml

```diff
@@ -6,21 +6,41 @@ metadata:
     app: keycloak
 spec:
   ports:
-    - name: http
+    - protocol: TCP
       port: 8080
-      targetPort: 8080
+      targetPort: http
+      name: http
   selector:
     app: keycloak
-  type: LoadBalancer
+  type: ClusterIP
+---
+apiVersion: v1
+kind: Service
+metadata:
+  labels:
+    app: keycloak
+  # Used to
+  name: keycloak-discovery
+spec:
+  selector:
+    app: keycloak
+  # Allow not-yet-ready Pods to be visible to ensure the forming of a cluster if Pods come up concurrently
+  publishNotReadyAddresses: true
+  clusterIP: None
+  type: ClusterIP
 ---
 apiVersion: apps/v1
-kind: Deployment
+# Use a stateful setup to ensure that for a rolling update Pods are restarted with a rolling strategy one-by-one.
+# This prevents losing in-memory information stored redundantly in two Pods.
+kind: StatefulSet
 metadata:
   name: keycloak
   labels:
     app: keycloak
 spec:
-  replicas: 1
+  serviceName: keycloak
```

Comment:
We're discussing this in the operator. The docs are not very consistent, but one statement from https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/ is that "StatefulSets currently require a [Headless Service](https://kubernetes.io/docs/concepts/services-networking/service/#headless-services) to be responsible for the network identity of the Pods. You are responsible for creating this Service."

So this should likely use the discovery service, not the main service.

cc @vmuzikar

Reply:
Addressed in next commit.